fs2/src/main/scala/laserdisc/fs2/RedisClient.scala

```diff
@@ -277,27 +262,29 @@ object RedisClient {
 
       val emptyState = State(hasShutdown = false, None)
 
-      async.refOf(emptyState).map { state =>
+      Ref.of(emptyState).map { state =>
         new Publisher[F] {
-          def shutdown: F[Unit] = state.modify(_.shutdown).flatMap {
-            case Change(p, _) =>
-              import cats.instances.option.catsStdInstancesForOption
-              p.maybeConnection.traverse_(_.shutdown)
-          }
+          val shutdown: F[Unit] =
+            state
+              .modify(currentState => (currentState.shutdown, currentState))
+              .flatMap(_.maybeConnection.traverse_(_.shutdown))
 
           def publish[In <: HList, Out <: HList](in: In, timeout: FiniteDuration)(
-              implicit protocolHandler: ProtocolHandler.Aux[F, In, Out]
+              implicit ev: ProtocolHandler.Aux[F, In, Out]
           ): F[Out] =
             state.get.map(_.maybeConnection).flatMap {
               case Some(connection) => connection.send(in, timeout)
               case None =>
                 createPublisher.flatMap { connection =>
                   state
-                    .modify(_.maybeSwapConnection(connection))
+                    .modify { currentState =>
+                      val newState = currentState.maybeSwapConnection(connection)
+                      (newState, (currentState, newState))
+                    }
                     .flatMap {
-                      case Change(p, _) if p.hasShutdown => F.raiseError[Out](ClientTerminated)
```

Comment:
I see you changed this in some places. This way the apply in `ConcurrentEffect`
```scala
def apply[F[_]](implicit instance : cats.effect.ConcurrentEffect[F]) : cats.effect.ConcurrentEffect[F]
```
will search the scope on every invocation. I know it's compile time, but could we find a compromise ? Something like a better name than F ?
Not the end of the world anyway.